platforms/jvm/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java

```diff
@@ -289,10 +289,11 @@ private static JavaVersion computeJavaTargetCompatibilityConvention(DefaultJavaP
     }
 
     private void configureScaladoc(final Project project, final ScalaRuntime scalaRuntime) {
+        ProviderFactory providers = project.getProviders();
         project.getTasks().withType(ScalaDoc.class).configureEach(scalaDoc -> {
-            scalaDoc.getConventionMapping().map("scalaClasspath", (Callable<FileCollection>) () -> scalaRuntime.inferScalaClasspath(scalaDoc.getClasspath()));
-            scalaDoc.getConventionMapping().map("destinationDir", (Callable<File>) () -> javaPluginExtension(project).getDocsDir().dir("scaladoc").get().getAsFile());
-            scalaDoc.getConventionMapping().map("title", (Callable<String>) () -> project.getExtensions().getByType(ReportingExtension.class).getApiDocTitle());
+            scalaDoc.getScalaClasspath().convention(providers.provider(() -> scalaRuntime.inferScalaClasspath(scalaDoc.getClasspath())));
```

Comment:
❓ Why do we need to wrap this in a `provider()`? I'd expect the `FileCollection` returned by `inferScalaClasspath()` should be good enough already, no? Constructing that object is not very costly, and it already provides the laziness we need here.